elftools/elf/dynamic.py

```diff
@@ -189,6 +250,39 @@ def __init__(self, header, stream, elffile):
                 break
         Segment.__init__(self, header, stream)
         Dynamic.__init__(self, stream, elffile, stringtable, self['p_offset'])
+        self._symbol_index_map = None
+        self._symbol_name_map = None
+        self._num_symbols = None
+
+    def num_symbols(self):
+        """ Number of symbols in the table recovered from DT_SYMTAB
+        """
+        if self._num_symbols is None:
+            self._num_symbols = len(list(self.iter_symbols()))
+        return self._num_symbols
+
+    def get_symbol(self, index):
+        """ Get the symbol at index #index from the table (Symbol object)
+        """
+        if self._symbol_index_map is None:
+            self._symbol_index_map = {}
+            for idx, symbol in enumerate(self.iter_symbols()):
+                self._symbol_index_map[idx] = symbol
+        return self._symbol_index_map[index]
+
+    def get_symbol_by_name(self, name):
+        """ Get a symbol(s) by name. Return None if no symbol by the given name
+            exists.
+        """
+        # The first time this method is called, construct a name to number
+        # mapping
+        #
+        if self._symbol_name_map is None:
+            self._symbol_name_map = defaultdict(list)
+            for i, sym in enumerate(self.iter_symbols()):
+                self._symbol_name_map[sym.name].append(i)
+        symnums = self._symbol_name_map.get(name)
+        return [self.get_symbol(i) for i in symnums] if symnums else None
```

Comment:
why None instead of an empty list?

Reply:
That's what SymbolTableSection does (https://github.com/eliben/pyelftools/blob/master/elftools/elf/sections.py#L174) so I figured this should behave the same way.